File: vibedev-specs-mcp/src/utils/project-analyzer.ts
import * as fs from 'fs';
import * as path from 'path';

export interface ProjectConfig {
  projectInfo: {
    name: string;
    analyzedAt: string;
    totalRepositories: number;
    totalServices: number;
  };
  repositoryMapping: Record<string, string>;
  serviceMapping: Record<string, ServiceInfo>;
  businessScenarios: BusinessScenario[];
  extractionPatterns: {
    traceIdPatterns: string[];
    userIdFields: string[];
  };
  databaseQueries: Record<string, QueryTemplates>;
}

export interface ServiceInfo {
  description: string;
  tables: string[];
  businessType: string;
  repositories: string[];
}

export interface BusinessScenario {
  scenario: string;
  relatedServices: string[];
  coreTables: string[];
  commonIssues: string[];
}

export interface QueryTemplates {
  basicQuery: string;
  countQuery: string;
  timeRangeQuery: string;
}

export class ProjectAnalyzer {
  private projectRoot: string;
  private repositoryMapping: Record<string, string> = {};
  private serviceMapping: Record<string, ServiceInfo> = {};
  private businessScenarios: BusinessScenario[] = [];
  private traceIdPatterns: string[] = [];
  private userIdFields: string[] = [];

  constructor(projectRoot: string = '.') {
    if (!projectRoot) {
      projectRoot = '.';
    }
    try {
      this.projectRoot = path.resolve(projectRoot);
      if (!this.projectRoot) {
        this.projectRoot = process.cwd();
      }
    } catch (error) {
      console.error(`Failed to resolve project root path: ${error}`);
      this.projectRoot = process.cwd();
    }
  }

  async analyzeProject(): Promise<ProjectConfig> {
    console.log('Starting project structure analysis...');
    
    // 1. Scan Repository classes
    await this.scanRepositories();
    
    // 2. Scan Service classes
    await this.scanServices();
    
    // 3. Analyze business scenarios
    this.analyzeBusinessScenarios();
    
    // 4. Extract TraceId patterns
    this.extractTracePatterns();
    
    // 5. Extract user identifier fields
    this.extractUserFields();
    
    return this.buildConfig();
  }

  private async scanRepositories(): Promise<void> {
    console.log('Scanning Repository classes...');
    
    const repoFiles = await this.findFiles('**/*Repository.java');
    
    for (const repoFile of repoFiles) {
      try {
        const content = await fs.promises.readFile(repoFile, 'utf-8');
        
        // Extract class name
        const classMatch = content.match(/class\s+(\w*Repository)/);
        if (classMatch) {
          const className = classMatch[1];
          
          // Infer table name
          const tableName = this.inferTableName(className, content);
          if (tableName) {
            this.repositoryMapping[className] = tableName;
          }
        }
      } catch (error) {
        console.error(`Failed to process Repository file ${repoFile}:`, error);
      }
    }
  }

  private inferTableName(className: string, content: string): string | null {
    // Method 1: Find @Table annotation
    const tableMatch = content.match(/@Table\s*\(\s*name\s*=\s*"([^"]+)"/);
    if (tableMatch) {
      return tableMatch[1];
    }
    
    // Method 2: Find table name in SQL
    const sqlPatterns = [
      /FROM\s+(\w+)/gi,
      /UPDATE\s+(\w+)/gi,
      /INSERT\s+INTO\s+(\w+)/gi,
      /DELETE\s+FROM\s+(\w+)/gi
    ];
    
    for (const pattern of sqlPatterns) {
      const matches = Array.from(content.matchAll(pattern));
      if (matches.length > 0) {
        // Return most common table name
        const tableNames = matches.map(m => m[1]);
        const counts = tableNames.reduce((acc, name) => {
          acc[name] = (acc[name] || 0) + 1;
          return acc;
        }, {} as Record<string, number>);
        
        return Object.keys(counts).reduce((a, b) => counts[a] > counts[b] ? a : b);
      }
    }
    
    // Method 3: Infer from class name (TpDealRepository -> tp_deal)
    if (className.endsWith('Repository')) {
      const baseName = className.slice(0, -10); // Remove Repository suffix
      // Convert camelCase to snake_case
      const tableName = baseName.replace(/([a-z0-9])([A-Z])/g, '$1_$2').toLowerCase();
      return tableName;
    }
    
    return null;
  }

  private async scanServices(): Promise<void> {
    console.log('Scanning Service classes...');
    
    const serviceFiles = await this.findFiles('**/*Service.java');
    
    for (const serviceFile of serviceFiles) {
      try {
        const content = await fs.promises.readFile(serviceFile, 'utf-8');
        
        // Extract class name
        const classMatch = content.match(/class\s+(\w*Service)/);
        if (classMatch) {
          const className = classMatch[1];
          
          // Analyze business function
          const serviceInfo = this.analyzeService(className, content);
          if (serviceInfo) {
            this.serviceMapping[className] = serviceInfo;
          }
        }
      } catch (error) {
        console.error(`Failed to process Service file ${serviceFile}:`, error);
      }
    }
  }

  private analyzeService(className: string, content: string): ServiceInfo | null {
    const serviceInfo: ServiceInfo = {
      description: '',
      tables: [],
      businessType: '',
      repositories: []
    };
    
    // Extract class comment as description
    const commentMatch = content.match(/\/\*\*\s*\n\s*\*\s*([^\n*]+)/);
    if (commentMatch) {
      serviceInfo.description = commentMatch[1].trim();
    }
    
    // Infer business type
    serviceInfo.businessType = this.inferBusinessType(className);
    
    // Find used Repository
    const repoMatches = Array.from(content.matchAll(/@Autowired[^;]*?(\w*Repository)/gs));
    serviceInfo.repositories = [...new Set(repoMatches.map(m => m[1]))];
    
    // Infer related tables based on Repository
    for (const repo of serviceInfo.repositories) {
      if (this.repositoryMapping[repo]) {
        serviceInfo.tables.push(this.repositoryMapping[repo]);
      }
    }
    
    return serviceInfo;
  }

  private inferBusinessType(className: string): string {
    const businessKeywords: Record<string, string> = {
      'Query': 'query',
      'Create': 'create',
      'Update': 'update',
      'Delete': 'delete',
      'Payment': 'payment',
      'Subs': 'deposit',
      'Holdings': 'holdings',
      'Trade': 'trade',
      'Order': 'order',
      'Validate': 'validate',
      'Migrate': 'migrate'
    };
    
    for (const [keyword, chinese] of Object.entries(businessKeywords)) {
      if (className.toLowerCase().includes(keyword.toLowerCase())) {
        return chinese;
      }
    }
    
    return 'unknown';
  }

  private analyzeBusinessScenarios(): void {
    console.log('Analyzing business scenarios...');
    
    // Analyze business scenarios based on Service grouping
    const scenarioGroups: Record<string, BusinessScenario> = {};
    
    for (const [serviceName, serviceInfo] of Object.entries(this.serviceMapping)) {
      const businessType = serviceInfo.businessType;
      
      if (!scenarioGroups[businessType]) {
        scenarioGroups[businessType] = {
          scenario: businessType,
          relatedServices: [],
          coreTables: [],
          commonIssues: []
        };
      }
      
      scenarioGroups[businessType].relatedServices.push(serviceName);
      scenarioGroups[businessType].coreTables.push(...serviceInfo.tables);
    }
    
    // Add common issues
    const issueMapping: Record<string, string[]> = {
      'payment': ['payment status error', 'amount calculation error', 'freeze/unfreeze failed'],
      'deposit': ['amount calculation error', 'transaction status error', 'share calculation error'],
      'query': ['data inconsistency', 'query timeout', 'permission validation failed'],
      'holdings': ['share calculation error', 'profit calculation error', 'data delay']
    };
    
    for (const [scenarioType, scenario] of Object.entries(scenarioGroups)) {
      scenario.coreTables = [...new Set(scenario.coreTables)];
      scenario.commonIssues = issueMapping[scenarioType] || ['unknown issue'];
      this.businessScenarios.push(scenario);
    }
  }

  private extractTracePatterns(): void {
    console.log('Extracting TraceId patterns...');
    
    // Common TraceId patterns
    this.traceIdPatterns = [
      'traceId[=:\\s]+([a-f0-9]{32})',
      'trace-id[=:\\s]+([a-f0-9-]{36})',
      'requestId[=:\\s]+([a-f0-9]{32})',
      'tid[=:\\s]+([a-f0-9]{32})',
      'X-Trace-Id[=:\\s]+([a-f0-9-]{36})'
    ];
  }

  private extractUserFields(): void {
    console.log('Extracting user identifier fields...');
    
    // Common user identifier fields
    this.userIdFields = [
      'custNo',
      'hboneNo',
      'customerId',
      'userId',
      'userNo',
      'clientId',
      'accountId'
    ];
  }

  private buildConfig(): ProjectConfig {
    return {
      projectInfo: {
        name: path.basename(this.projectRoot),
        analyzedAt: new Date().toISOString(),
        totalRepositories: Object.keys(this.repositoryMapping).length,
        totalServices: Object.keys(this.serviceMapping).length
      },
      repositoryMapping: this.repositoryMapping,
      serviceMapping: this.serviceMapping,
      businessScenarios: this.businessScenarios,
      extractionPatterns: {
        traceIdPatterns: this.traceIdPatterns,
        userIdFields: this.userIdFields
      },
      databaseQueries: this.generateQueryTemplates()
    };
  }

  private generateQueryTemplates(): Record<string, QueryTemplates> {
    const templates: Record<string, QueryTemplates> = {};
    
    // Generate basic query templates for each table
    for (const [repo, table] of Object.entries(this.repositoryMapping)) {
      templates[table] = {
        basicQuery: `SELECT * FROM ${table} WHERE {condition} LIMIT 10`,
        countQuery: `SELECT COUNT(*) FROM ${table} WHERE {condition}`,
        timeRangeQuery: `SELECT * FROM ${table} WHERE created_time >= '{start_time}' AND created_time <= '{end_time}' LIMIT 20`
      };
    }
    
    return templates;
  }

  async saveConfig(outputPath?: string): Promise<ProjectConfig> {
    if (!outputPath) {
      outputPath = path.join(this.projectRoot, 'bugfix.project.auto.json');
    }
    
    const config = this.buildConfig();
    
    // Create directory if not exists
    const outputDir = path.dirname(outputPath);
    if (outputDir && outputDir !== '.') {
      await fs.promises.mkdir(outputDir, { recursive: true });
    }
    
    await fs.promises.writeFile(outputPath, JSON.stringify(config, null, 2), 'utf-8');
    
    console.log(`Project configuration saved to: ${outputPath}`);
    return config;
  }

  private async findFiles(pattern: string): Promise<string[]> {
    const files: string[] = [];
    
    const walkDir = async (dir: string): Promise<void> => {
      try {
        const entries = await fs.promises.readdir(dir, { withFileTypes: true });
        
        for (const entry of entries) {
          const fullPath = path.join(dir, entry.name);
          
          if (entry.isDirectory()) {
            await walkDir(fullPath);
          } else if (entry.isFile()) {
            // Simple pattern matching for Java files
            if (pattern.includes('Repository.java') && entry.name.endsWith('Repository.java')) {
              files.push(fullPath);
            } else if (pattern.includes('Service.java') && entry.name.endsWith('Service.java')) {
              files.push(fullPath);
            }
          }
        }
      } catch (error) {
        // Ignore directories we can't read
      }
    };
    
    await walkDir(this.projectRoot);
    return files;
  }
}

File: .github/chatmodes/project_analyzer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Project Auto Analyzer - Auto identify project structure and business mappings
"""
import os
import re
import json
import sys
import glob

class ProjectAnalyzer:
    """Project structure auto analyzer"""
    
    def __init__(self, project_root):
        """Initialize project analyzer
        
        Args:
            project_root: Project root directory path
        """
        self.project_root = os.path.abspath(project_root)
        self.repository_mapping = {}
        self.service_mapping = {}
        self.business_scenarios = []
        self.trace_id_patterns = []
        self.user_id_fields = []
        
    def analyze_project(self):
        """Analyze entire project structure
        
        Returns:
            Project analysis result dictionary
        """
        print("Starting project structure analysis...")
        
        # 1. Scan Repository classes
        self._scan_repositories()
        
        # 2. Scan Service classes
        self._scan_services()
        
        # 3. Analyze business scenarios
        self._analyze_business_scenarios()
        
        # 4. Extract TraceId patterns
        self._extract_trace_patterns()
        
        # 5. Extract user identifier fields
        self._extract_user_fields()
        
        return self._build_config()
    
    def _scan_repositories(self):
        """Scan Repository classes and infer table names"""
        print("Scanning Repository classes...")
        
        # Find all Repository files
        repo_files = []
        for root, dirs, files in os.walk(self.project_root):
            for file in files:
                if file.endswith('Repository.java'):
                    repo_files.append(os.path.join(root, file))
        
        for repo_file in repo_files:
            try:
                with open(repo_file, 'r') as f:
                    content = f.read()
                
                # Extract class name
                class_match = re.search(r'class\s+(\w*Repository)', content)
                if class_match:
                    class_name = class_match.group(1)
                    
                    # Infer table name
                    table_name = self._infer_table_name(class_name, content)
                    if table_name:
                        self.repository_mapping[class_name] = table_name
                        
            except Exception as e:
                print("Failed to process Repository file %s: %s" % (repo_file, str(e)))
    
    def _infer_table_name(self, class_name, content):
        """Infer table name from Repository class
        
        Args:
            class_name: Repository class name
            content: File content
            
        Returns:
            Inferred table name
        """
        # Method 1: Find @Table annotation
        table_match = re.search(r'@Table\s*\(\s*name\s*=\s*"([^"]+)"', content)
        if table_match:
            return table_match.group(1)
        
        # Method 2: Find table name in SQL
        sql_patterns = [
            r'FROM\s+(\w+)',
            r'UPDATE\s+(\w+)',
            r'INSERT\s+INTO\s+(\w+)',
            r'DELETE\s+FROM\s+(\w+)'
        ]
        
        for pattern in sql_patterns:
            matches = re.findall(pattern, content, re.IGNORECASE)
            if matches:
                # Return most common table name
                return max(set(matches), key=matches.count)
        
        # Method 3: Infer from class name (TpDealRepository -> tp_deal)
        if class_name.endswith('Repository'):
            base_name = class_name[:-10]  # Remove Repository suffix
            # Convert camelCase to snake_case
            table_name = re.sub(r'([a-z0-9])([A-Z])', r'\1_\2', base_name).lower()
            return table_name
        
        return None
    
    def _scan_services(self):
        """Scan Service classes and analyze business functions"""
        print("Scanning Service classes...")
        
        # Find all Service files
        service_files = []
        for root, dirs, files in os.walk(self.project_root):
            for file in files:
                if file.endswith('Service.java'):
                    service_files.append(os.path.join(root, file))
        
        for service_file in service_files:
            try:
                with open(service_file, 'r') as f:
                    content = f.read()
                
                # Extract class name
                class_match = re.search(r'class\s+(\w*Service)', content)
                if class_match:
                    class_name = class_match.group(1)
                    
                    # Analyze business function
                    service_info = self._analyze_service(class_name, content)
                    if service_info:
                        self.service_mapping[class_name] = service_info
                        
            except Exception as e:
                print("Failed to process Service file %s: %s" % (service_file, str(e)))
    
    def _analyze_service(self, class_name, content):
        """Analyze Service class business function
        
        Args:
            class_name: Service class name
            content: File content
            
        Returns:
            Service information dictionary
        """
        service_info = {
            "description": "",
            "tables": [],
            "businessType": "",
            "repositories": []
        }
        
        # Extract class comment as description
        comment_match = re.search(r'/\*\*\s*\n\s*\*\s*([^\n*]+)', content)
        if comment_match:
            service_info["description"] = comment_match.group(1).strip()
        
        # Infer business type
        service_info["businessType"] = self._infer_business_type(class_name)
        
        # Find used Repository
        repo_matches = re.findall(r'@Autowired[^;]*?(\w*Repository)', content, re.DOTALL)
        service_info["repositories"] = list(set(repo_matches))
        
        # Infer related tables based on Repository
        for repo in service_info["repositories"]:
            if repo in self.repository_mapping:
                service_info["tables"].append(self.repository_mapping[repo])
        
        return service_info
    
    def _infer_business_type(self, class_name):
        """Infer business type from class name
        
        Args:
            class_name: Service class name
            
        Returns:
            Business type
        """
        business_keywords = {
            "Query": "query",
            "Create": "create",
            "Update": "update",
            "Delete": "delete",
            "Payment": "payment",
            "Subs": "deposit",
            "Holdings": "holdings",
            "Trade": "trade",
            "Order": "order",
            "Validate": "validate",
            "Migrate": "migrate"
        }
        
        for keyword, chinese in business_keywords.items():
            if keyword.lower() in class_name.lower():
                return chinese
        
        return "unknown"
    
    def _analyze_business_scenarios(self):
        """Analyze business scenarios"""
        print("Analyzing business scenarios...")
        
        # Analyze business scenarios based on Service grouping
        scenario_groups = {}
        
        for service_name, service_info in self.service_mapping.items():
            business_type = service_info["businessType"]
            
            if business_type not in scenario_groups:
                scenario_groups[business_type] = {
                    "scenario": business_type,
                    "relatedServices": [],
                    "coreTables": set(),
                    "commonIssues": []
                }
            
            scenario_groups[business_type]["relatedServices"].append(service_name)
            scenario_groups[business_type]["coreTables"].update(service_info["tables"])
        
        # Add common issues
        issue_mapping = {
            "payment": ["payment status error", "amount calculation error", "freeze/unfreeze failed"],
            "deposit": ["amount calculation error", "transaction status error", "share calculation error"],
            "query": ["data inconsistency", "query timeout", "permission validation failed"],
            "holdings": ["share calculation error", "profit calculation error", "data delay"]
        }
        
        for scenario_type, scenario_info in scenario_groups.items():
            scenario_info["coreTables"] = list(scenario_info["coreTables"])
            scenario_info["commonIssues"] = issue_mapping.get(scenario_type, ["unknown issue"])
            self.business_scenarios.append(scenario_info)
    
    def _extract_trace_patterns(self):
        """Extract TraceId patterns"""
        print("Extracting TraceId patterns...")
        
        # Common TraceId patterns
        common_patterns = [
            r'traceId[=:\s]+([a-f0-9]{32})',
            r'trace-id[=:\s]+([a-f0-9-]{36})',
            r'requestId[=:\s]+([a-f0-9]{32})',
            r'tid[=:\s]+([a-f0-9]{32})',
            r'X-Trace-Id[=:\s]+([a-f0-9-]{36})'
        ]
        
        self.trace_id_patterns = common_patterns
    
    def _extract_user_fields(self):
        """Extract user identifier fields"""
        print("Extracting user identifier fields...")
        
        # Common user identifier fields
        common_fields = [
            "custNo",
            "hboneNo",
            "customerId", 
            "userId",
            "userNo",
            "clientId",
            "accountId"
        ]
        
        self.user_id_fields = common_fields
    
    def _build_config(self):
        """Build configuration dictionary
        
        Returns:
            Complete project configuration
        """
        config = {
            "projectInfo": {
                "name": os.path.basename(self.project_root),
                "analyzedAt": "auto-generated",
                "totalRepositories": len(self.repository_mapping),
                "totalServices": len(self.service_mapping)
            },
            "repositoryMapping": self.repository_mapping,
            "serviceMapping": self.service_mapping,
            "businessScenarios": self.business_scenarios,
            "extractionPatterns": {
                "traceIdPatterns": self.trace_id_patterns,
                "userIdFields": self.user_id_fields
            },
            "databaseQueries": self._generate_query_templates()
        }
        
        return config
    
    def _generate_query_templates(self):
        """Generate query templates
        
        Returns:
            Query template dictionary
        """
        templates = {}
        
        # Generate basic query templates for each table
        for repo, table in self.repository_mapping.items():
            templates[table] = {
                "basicQuery": "SELECT * FROM %s WHERE {condition} LIMIT 10" % table,
                "countQuery": "SELECT COUNT(*) FROM %s WHERE {condition}" % table,
                "timeRangeQuery": "SELECT * FROM %s WHERE created_time >= '{start_time}' AND created_time <= '{end_time}' LIMIT 20" % table
            }
        
        return templates
    
    def save_config(self, output_path=None):
        """Save configuration to file
        
        Args:
            output_path: Output file path
        """
        if output_path is None:
            output_path = os.path.join(self.project_root, ".github", "chatmodes", "bugfix.project.auto.json")
        
        config = self._build_config()
        
        # Create directory if not exists
        output_dir = os.path.dirname(output_path)
        if not os.path.exists(output_dir):
            os.makedirs(output_dir)
        
        with open(output_path, 'w') as f:
            json.dump(config, f, ensure_ascii=False, indent=2)
        
        print("Project configuration saved to: %s" % output_path)
        return config

def main():
    """Main function"""
    if len(sys.argv) < 2:
        project_root = os.getcwd()
    else:
        project_root = sys.argv[1]
    
    analyzer = ProjectAnalyzer(project_root)
    config = analyzer.analyze_project()
    analyzer.save_config()
    
    print("\nAnalysis completed:")
    print("- Repository classes: %d" % len(config['repositoryMapping']))
    print("- Service classes: %d" % len(config['serviceMapping']))
    print("- Business scenarios: %d" % len(config['businessScenarios']))

if __name__ == "__main__":
    main()
